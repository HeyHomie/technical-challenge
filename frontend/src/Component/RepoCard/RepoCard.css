.repo-link {
  color: #0a58ca;
  text-decoration: none;
  font-size: 20px;
  font-weight: 600;
}

.repo-description {
  color: #8b949e;
}

.repo-topic {
  color: #58a6ff;
  background-color: rgba(56, 139, 253, 0.15);
  border-radius: 2em;
  padding-right: 10px;
  padding-left: 10px;
  line-height: 22px;
  text-decoration: none;
  margin-right: 4px;
}

.repo-label {
  color: 5f666d;
  display: inline-block;
  padding: 0 7px;
  font-size: 12px;
  font-weight: 500;
  line-height: 18px;
  white-space: nowrap;
  border: 1px solid transparent;
  border-radius: 2em;
  border-color: #ced4db;
  margin-left: 5px;
  vertical-align: middle;
  text-transform: capitalize;
}
.repo-info {
  display: inline-flex;
  font-size: 14px;
}
.repo-info_item {
  margin-right: 15px;
}

.repo-info_languageColor {
  position: relative;
  top: 1px;
  display: inline-block;
  width: 12px;
  height: 12px;
  /* border: 1px solid var(--color-primer-border-contrast); */
  border-radius: 50%;
  margin-right: 5px;
}
