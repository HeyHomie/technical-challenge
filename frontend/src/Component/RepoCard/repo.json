{
  "id": 59423583,
  "node_id": "MDEwOlJlcG9zaXRvcnk1OTQyMzU4Mw==",
  "name": "163-music-unlock",
  "full_name": "yknx4/163-music-unlock",
  "private": false,
  "owner": {
    "login": "yknx4",
    "id": 1848186,
    "node_id": "MDQ6VXNlcjE4NDgxODY=",
    "avatar_url": "https://avatars.githubusercontent.com/u/1848186?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/yknx4",
    "html_url": "https://github.com/yknx4",
    "followers_url": "https://api.github.com/users/yknx4/followers",
    "following_url": "https://api.github.com/users/yknx4/following{/other_user}",
    "gists_url": "https://api.github.com/users/yknx4/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/yknx4/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/yknx4/subscriptions",
    "organizations_url": "https://api.github.com/users/yknx4/orgs",
    "repos_url": "https://api.github.com/users/yknx4/repos",
    "events_url": "https://api.github.com/users/yknx4/events{/privacy}",
    "received_events_url": "https://api.github.com/users/yknx4/received_events",
    "type": "User",
    "site_admin": false
  },
  "html_url": "https://github.com/yknx4/163-music-unlock",
  "description": "[已更新] 网易云音乐移动客户端反代，解除海外限制，可播放下架歌曲和付费歌曲",
  "fork": true,
  "url": "https://api.github.com/repos/yknx4/163-music-unlock",
  "forks_url": "https://api.github.com/repos/yknx4/163-music-unlock/forks",
  "keys_url": "https://api.github.com/repos/yknx4/163-music-unlock/keys{/key_id}",
  "collaborators_url": "https://api.github.com/repos/yknx4/163-music-unlock/collaborators{/collaborator}",
  "teams_url": "https://api.github.com/repos/yknx4/163-music-unlock/teams",
  "hooks_url": "https://api.github.com/repos/yknx4/163-music-unlock/hooks",
  "issue_events_url": "https://api.github.com/repos/yknx4/163-music-unlock/issues/events{/number}",
  "events_url": "https://api.github.com/repos/yknx4/163-music-unlock/events",
  "assignees_url": "https://api.github.com/repos/yknx4/163-music-unlock/assignees{/user}",
  "branches_url": "https://api.github.com/repos/yknx4/163-music-unlock/branches{/branch}",
  "tags_url": "https://api.github.com/repos/yknx4/163-music-unlock/tags",
  "blobs_url": "https://api.github.com/repos/yknx4/163-music-unlock/git/blobs{/sha}",
  "git_tags_url": "https://api.github.com/repos/yknx4/163-music-unlock/git/tags{/sha}",
  "git_refs_url": "https://api.github.com/repos/yknx4/163-music-unlock/git/refs{/sha}",
  "trees_url": "https://api.github.com/repos/yknx4/163-music-unlock/git/trees{/sha}",
  "statuses_url": "https://api.github.com/repos/yknx4/163-music-unlock/statuses/{sha}",
  "languages_url": "https://api.github.com/repos/yknx4/163-music-unlock/languages",
  "stargazers_url": "https://api.github.com/repos/yknx4/163-music-unlock/stargazers",
  "contributors_url": "https://api.github.com/repos/yknx4/163-music-unlock/contributors",
  "subscribers_url": "https://api.github.com/repos/yknx4/163-music-unlock/subscribers",
  "subscription_url": "https://api.github.com/repos/yknx4/163-music-unlock/subscription",
  "commits_url": "https://api.github.com/repos/yknx4/163-music-unlock/commits{/sha}",
  "git_commits_url": "https://api.github.com/repos/yknx4/163-music-unlock/git/commits{/sha}",
  "comments_url": "https://api.github.com/repos/yknx4/163-music-unlock/comments{/number}",
  "issue_comment_url": "https://api.github.com/repos/yknx4/163-music-unlock/issues/comments{/number}",
  "contents_url": "https://api.github.com/repos/yknx4/163-music-unlock/contents/{+path}",
  "compare_url": "https://api.github.com/repos/yknx4/163-music-unlock/compare/{base}...{head}",
  "merges_url": "https://api.github.com/repos/yknx4/163-music-unlock/merges",
  "archive_url": "https://api.github.com/repos/yknx4/163-music-unlock/{archive_format}{/ref}",
  "downloads_url": "https://api.github.com/repos/yknx4/163-music-unlock/downloads",
  "issues_url": "https://api.github.com/repos/yknx4/163-music-unlock/issues{/number}",
  "pulls_url": "https://api.github.com/repos/yknx4/163-music-unlock/pulls{/number}",
  "milestones_url": "https://api.github.com/repos/yknx4/163-music-unlock/milestones{/number}",
  "notifications_url": "https://api.github.com/repos/yknx4/163-music-unlock/notifications{?since,all,participating}",
  "labels_url": "https://api.github.com/repos/yknx4/163-music-unlock/labels{/name}",
  "releases_url": "https://api.github.com/repos/yknx4/163-music-unlock/releases{/id}",
  "deployments_url": "https://api.github.com/repos/yknx4/163-music-unlock/deployments",
  "created_at": "2016-05-22T16:55:42Z",
  "updated_at": "2019-04-17T16:44:43Z",
  "pushed_at": "2021-08-19T14:49:59Z",
  "git_url": "git://github.com/yknx4/163-music-unlock.git",
  "ssh_url": "git@github.com:yknx4/163-music-unlock.git",
  "clone_url": "https://github.com/yknx4/163-music-unlock.git",
  "svn_url": "https://github.com/yknx4/163-music-unlock",
  "homepage": "",
  "size": 8344,
  "stargazers_count": 0,
  "watchers_count": 0,
  "language": "Python",
  "has_issues": false,
  "has_projects": true,
  "has_downloads": true,
  "has_wiki": true,
  "has_pages": false,
  "forks_count": 0,
  "mirror_url": null,
  "archived": false,
  "disabled": false,
  "open_issues_count": 1,
  "license": null,
  "allow_forking": true,
  "is_template": false,
  "topics": [],
  "visibility": "public",
  "forks": 0,
  "open_issues": 1,
  "watchers": 0,
  "default_branch": "master"
}
