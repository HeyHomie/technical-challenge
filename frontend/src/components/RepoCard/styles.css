.RepoCard--container{
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  border-top: 1px solid var(--divider-color);
  margin: 0 20px;
  padding: 30px 0 20px;
  
}

/* Header */

.RepoCard__header{
  display: flex;
  margin-bottom: 10px;
}

.RepoCard__header h2{
  color: var(--font-link-color);
  font-weight: 500;
}

.RepoCard__header p{
  align-self: center;
  margin-left: 20px;
  color: gray;
  border: 1px solid gray;
  border-radius: 20px 20px;
  padding: 1px 8px;
  font-size: 0.8rem;
}

/* Description */

.RepoCard__description{
  margin-bottom: 10px;
  font-size: 0.9rem;
  
}

.RepoCard__description p {
  color: gray;
  margin-bottom: 10px;
}

.RepoCard__description--topics{
  display: inline-block;
  color: #58a6ff;
  border-radius: 20px 20px;
  background-color: #092353;
  padding: 2px 10px;
  margin-right: 3px;
  margin-bottom: 5px;
}

.RepoCard__description--topics:hover{
  background-color: #1F6FEB;
}

/* Info */

.RepoCard__info{
  display: flex;
  align-items: flex-start;
  margin-bottom: 10px;
  font-size: 0.9rem
}

.RepoCard__info p{
  color: gray;
  margin-right: 10px;
}