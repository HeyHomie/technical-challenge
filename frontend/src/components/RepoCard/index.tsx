import React, { FunctionComponent } from 'react'
import { UpdatedAtFormated } from '../updatedAtFormated'
import './styles.css'
interface Props {
  name: string
  private: boolean
  isPrivate: boolean
  description?: string
  topics: string[]
  language?: string
  fork: boolean
  forks: number
  license?: any
  updated_at: string
  updatedAt: string
  html_url: string
  htmlUrl: string
}

export const RepoCard: FunctionComponent<Props> = ({
  name,
  private: isPrivate,
  description,
  topics,
  language,
  fork,
  forks,
  license,
  updated_at: updatedAt,
  html_url: htmlUrl
}) => {
  let id: number = 0

  return (
    <div className='RepoCard--container'>
      <div className='RepoCard__header'>
        <a href={htmlUrl}>
          <h2>{name}</h2>
        </a>
        {!isPrivate && <p>Public</p>}
      </div>

      <div className='RepoCard__description'>
        <p>{description}</p>
        {topics.length > 0 &&
          topics.map((item) => (
            <span key={(id = id + 1)} className='RepoCard__description--topics'>
              {item}
            </span>
          ))}
      </div>

      <div className='RepoCard__info'>
        {language != null && <p>{language}</p>}
        {/* {fork && <a href="">{forks}</a>} */} {/* Fetching error */}
        {license != null && <p>{license.name}</p>}
        <UpdatedAtFormated timeDate={updatedAt} />
      </div>
    </div>
  )
}
