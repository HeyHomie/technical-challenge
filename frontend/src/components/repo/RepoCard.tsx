import { FC } from 'react'
import { IRepository, Language } from 'types'
import { Tag } from 'components/ui'
import { LawIcon, StarIcon } from '@primer/octicons-react'

const colors = {
  [Language.TypeScript]: '#2b7489',
  [Language.JavaScript]: '#f1e05a',
  [Language.HTML]: '#e34c26',
  [Language.CSS]: '#563d7c',
  [Language.Elixir]: '#6e4a7e',
  [Language.Java]: '#b07219',
  [Language.Python]: '#3572A5',
  [Language.Ruby]: '#701516',
  [Language.Go]: '#00ADD9',
  [Language.Rust]: '#dea584',
  [Language.Swift]: '#ffac45',
  [Language.Kotlin]: '#f18e33',
  [Language.C]: '#555555',
  [Language['C#']]: '#178600',
  [Language['C++']]: '#f34b7d',
  [Language.TeX]: '#3D6117',
  [Language.Shell]: '#89e051',
  [Language['Objective-C']]: '#438eff'
}

type RepoCardProps = {
  repo: IRepository
}
const RepoCard: FC<RepoCardProps> = ({ repo }) => {
  const diffDates = (date1: string, date2: string | number) => {
    const date1_ms = new Date(date1).getTime()
    const date2_ms = new Date(date2).getTime()
    const diff_ms = date2_ms - date1_ms

    return diff_ms
  }

  const convertDate = (date: string) => {
    const seconds = Math.floor(diffDates(date, Date.now()) / 1000)
    const hours = Math.floor(seconds / 60 / 60)
    const days = Math.floor(hours / 24)
    const months = Math.floor(days / 30)
    const years = Math.floor(months / 12)

    const days_text = days > 1 ? 'days' : 'day'
    const hours_text = hours > 1 ? 'hours' : 'hour'
    const months_text = months > 1 ? 'months' : 'month'
    const years_text = years > 1 ? 'years' : 'year'

    const time_text =
      years > 0
        ? `${years} ${years_text} ago`
        : months > 0
        ? `${months} ${months_text} ago`
        : days > 0
        ? `${days} ${days_text} ago`
        : hours > 0
        ? `${hours} ${hours_text} ago`
        : 'less than a minute'

    return time_text
  }

  const updatedAt = convertDate(repo.updated_at)

  return (
    <div className="grid grid-cols-6 py-6">
      <div className="grid col-span-5 gap-2 ">
        <h3>
          <a
            href={repo.html_url}
            className="mr-2 text-xl font-bold text-accent-fg">
            {repo.name}
          </a>
          <Tag className="text-secondary">
            {repo.private ? 'Private' : 'Public'}
          </Tag>
        </h3>
        <p className="text-secondary">{repo.description}</p>
        <div className="flex gap-4 ">
          {repo.topics.map((topic) => (
            <Tag className="bg-accent-subtle hover:bg-accent-emphasis text-accent-fg">
              {topic}
            </Tag>
          ))}
        </div>
        <div className="inline-flex gap-4 text-secondary">
          {repo.language ? (
            <p>
              <span
                className="inline-block w-3 h-3 mr-1 rounded-full"
                style={{ backgroundColor: colors[repo.language] }}></span>
              {repo.language}
            </p>
          ) : null}
          {repo.license ? (
            <p>
              {repo.license?.key === 'mit' ? <LawIcon className="mr-1" /> : ''}
              {repo.license?.name}
            </p>
          ) : null}
          <p>Updated {updatedAt}</p>
        </div>
      </div>
      <div className="col-span-1">
        <button className="w-auto btn">
          <StarIcon className="text-secondary" /> Star
        </button>
      </div>
    </div>
  )
}

export default RepoCard
