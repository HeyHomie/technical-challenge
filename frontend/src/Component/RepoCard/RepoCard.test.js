import { render, screen } from '@testing-library/react'
import RepoCard from './RepoCard'
import repoJSON from './repo.json'
import { BrowserRouter } from 'react-router-dom'

test('Render props', () => {
  render(
    <BrowserRouter>
      <RepoCard repo={repoJSON} username="Jorge" />
    </BrowserRouter>
  )
})

describe('Component Mounted', () => {
  it('renders the component', () => {
    const { container } = render(
      <BrowserRouter>
        <RepoCard repo={repoJSON} username="Jorge" />
      </BrowserRouter>
    )
    screen.logTestingPlawygroundURL()
    expect(container.firstChild).toMatchSnapshot()

    screen.getByText(
      /ðŸ“ˆ uptime monitor and status page for ale ornelas figueroa, powered by @upptime/i
    )
  })
})
