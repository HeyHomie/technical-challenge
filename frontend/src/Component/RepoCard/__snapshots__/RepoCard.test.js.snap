// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Mounted renders the component 1`] = `
<div
  class="mb-3 container"
>
  <div
    class="row"
  >
    <hr />
  </div>
  <div
    class="row"
  >
    <div>
      <h3>
        <a
          class="repo-link"
          href="https://github.com/Jorge/aleworld-status"
          rel="noopener noreferrer"
          target="_blank"
        >
          aleworld-status
        </a>
        <span
          class="repo-label mb-1"
        >
          public
        </span>
      </h3>
    </div>
  </div>
  <div
    class="row"
  >
    <p
      class="repo-description"
    >
      ðŸ“ˆ Uptime monitor and status page for Ale Ornelas Figueroa, powered by @upptime
    </p>
  </div>
  <div
    class="row"
  >
    <div
      class="my-1 d-inline-flex"
    >
      <a
        class="repo-topic"
        href="https://github.com/topics/status-page"
        rel="noopener noreferrer"
        target="_blank"
      >
        status-page
      </a>
      <a
        class="repo-topic"
        href="https://github.com/topics/upptime"
        rel="noopener noreferrer"
        target="_blank"
      >
        upptime
      </a>
      <a
        class="repo-topic"
        href="https://github.com/topics/uptime-monitor"
        rel="noopener noreferrer"
        target="_blank"
      >
        uptime-monitor
      </a>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="repo-info"
    >
      <div
        class="repo-info_item"
      >
        <span>
          <svg
            aria-hidden="true"
            class="octicon octicon-law mr-1"
            data-view-component="true"
            height="16"
            version="1.1"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M8.75.75a.75.75 0 00-1.5 0V2h-.984c-.305 0-.604.08-.869.23l-1.288.737A.25.25 0 013.984 3H1.75a.75.75 0 000 1.5h.428L.066 9.192a.75.75 0 00.154.838l.53-.53-.53.53v.001l.002.002.002.002.006.006.016.015.045.04a3.514 3.514 0 00.686.45A4.492 4.492 0 003 11c.88 0 1.556-.22 2.023-.454a3.515 3.515 0 00.686-.45l.045-.04.016-.015.006-.006.002-.002.001-.002L5.25 9.5l.53.53a.75.75 0 00.154-.838L3.822 4.5h.162c.305 0 .604-.08.869-.23l1.289-.737a.25.25 0 01.124-.033h.984V13h-2.5a.75.75 0 000 1.5h6.5a.75.75 0 000-1.5h-2.5V3.5h.984a.25.25 0 01.124.033l1.29.736c.264.152.563.231.868.231h.162l-2.112 4.692a.75.75 0 00.154.838l.53-.53-.53.53v.001l.002.002.002.002.006.006.016.015.045.04a3.517 3.517 0 00.686.45A4.492 4.492 0 0013 11c.88 0 1.556-.22 2.023-.454a3.512 3.512 0 00.686-.45l.045-.04.01-.01.006-.005.006-.006.002-.002.001-.002-.529-.531.53.53a.75.75 0 00.154-.838L13.823 4.5h.427a.75.75 0 000-1.5h-2.234a.25.25 0 01-.124-.033l-1.29-.736A1.75 1.75 0 009.735 2H8.75V.75zM1.695 9.227c.285.135.718.273 1.305.273s1.02-.138 1.305-.273L3 6.327l-1.305 2.9zm10 0c.285.135.718.273 1.305.273s1.02-.138 1.305-.273L13 6.327l-1.305 2.9z"
              fill-rule="evenodd"
            />
          </svg>
           
          MIT License
        </span>
      </div>
      <div>
        Updated on 24 May 2022 
      </div>
    </div>
  </div>
  <div
    class="row"
  />
</div>
`;
